### PR TITLE
feat: adjust  FA3 num_splits on Hopper for low latency in cudagraph mode

### PR DESCRIPTION
**Pull request by jmkuebler**
## Purpose

FA3 parallelizes attention across batch size and KV heads naturally. However, parallelization across the context dimension (splitting the KV sequence into chunks, similar to context-parallel decoding) incurs overhead from the combine kernel that merges partial results. vLLM currently by default uses `flash_attn_max_num_splits_for_cuda_graph = 32` when capturing CUDA graphs for decode.

When the batch size is small and only one or a few KV heads reside on each GPU (e.g. after tensor parallelism), most SMs sit idle. For example, with 1 KV head per device and batch size 1 on an H100 (132 SMs), only 1 CTA is launched without splits — the default of 32 splits still leaves 100 SMs unused.

This PR adds a heuristic that automatically increases `num_splits` to fill all SMs when the model is provably underutilizing them. The change is very conservative:

- Only activates on Hopper (SM90) with FA3
- Only activates when `max_model_len > 100k` (avoids combine kernel overhead regression at short context)
- Only activates when the user hasn't explicitly set the value
- Never goes below the original default of 32, capped at `min(num_sms, 256)`

For most use cases the heuristic falls back to 32 (for example if the max number of sequences is >= 4). The change is somewhat forward-looking, as models trend toward longer context and fewer KV heads (GQA/MQA), this becomes increasingly relevant. And it is then particularly relevant in an extreme latency optimized setting where one runs at BS=1 and with many GPUs (such that only one KV head per GPU).

[For context, check Section B.9 of the FA3 paper [https://openreview.net/pdf?id=tVConYid20](https://openreview.net/pdf?id=tVConYid20)]

### Memory impact

When `num_splits > 1`, FA3 allocates intermediate buffers to store partial attention
results before the combine kernel reduces them. To my understanding, during CUDA graph capture, these
allocations become permanent, since each captured graph retains its own buffers.

However, since our heuristic only kicks in when the `max_num_seqs` is small, in this case also the captured sizes get reduced and overall the impact on memory footprint is tiny and negligible. 
We also tried to read it from the startup logs, but ran into a small bug https://github.com/vllm-project/vllm/issues/35983 

**Formula**
Per captured batch size `B`:

```
split_buffer(B) = num_splits × num_q_heads × B × head_dim × sizeof(dtype)
               + num_splits × num_q_heads × B × sizeof(float32)   [LSE accum]
```

The LSE (log-sum-exp) accumulator is small relative to the output buffer, so approximately:

```
split_buffer(B) ≈ num_splits × num_q_heads × B × head_dim × sizeof(dtype)
```

Total across all captured graphs:

```
total = num_splits × num_q_heads × head_dim × sizeof(dtype) × Σ B_i
        for each B_i in cudagraph_capture_sizes
```

**Default capture sizes**

With default `max_num_seqs`, vLLM captures graphs for batch sizes `[1, 2, 4, 8, 16, 24, 32, ..., 248, 256]`, totaling `Σ B_i = 4,231`. Capture sizes are capped by `max_num_seqs`, so with `max_num_seqs=1` only `[1]` is captured (`Σ B_i = 1`).

**Example: Llama 3.1 8B (TP=1, 32 query heads)**

Our heuristic does not change `num_splits` here. Calculations are for reference.

Per split: `32 × 128 × 2 × 4,231 = 34.7 MB`

| num_splits | Buffer memory | Note |
|------------|--------------|------|
| 1          | 0 MB | no intermediate buffer |
| 16         | 555 MB | |
| 32         | 1,109 MB | ← current default, **our heuristic** |
| 132        | 4,575 MB | |

**Example: Llama 3.1 8B (TP=8, 4 query heads/device)**

**`max_num_seqs=1`** → `Σ B_i = 1`. Per split: `4 × 128 × 2 × 1 = 1 KB`

| num_splits | Buffer memory | Note |
|------------|--------------|------|
| 32         | 33 KB | ← current default |
| 132        | 135 KB | ← **our heuristic** (+102 KB) |

**`max_num_seqs=default`** → `Σ B_i = 4,231`. Per split: `4 × 128 × 2 × 4,231 = 4.3 MB`

| num_splits | Buffer memory | Note |
|------------|--------------|------|
| 32         | 139 MB | ← current default, **our heuristic** |
| 132        | 572 MB | (not triggered by heuristic) |

The heuristic only increases splits when `max_num_seqs` is small, which is exactly when few graphs are captured and memory impact is minimal.

## Test Plan

1. Unit tests for `compute_fa3_num_splits` (pure math, 7 parametrized cases covering floor/cap/edge cases)
2. Integration tests for `_resolve_fa_num_splits` with mocked platform (5 cases: Hopper+FA3, many KV heads, FA2 fallback, non-CUDA fallback, short-context fallback)
3. Benchmarks on p5.48xlarge (8× H100 SXM) with `vllm bench serve`, random dataset, `output_len=128`, `request_rate=inf`:
   - Llama 3.1 8B (TP=8, 1 KV head/device) — config modified to simulate long context beyond native 128k
   - Qwen3.5-35B-A3B MoE (TP=2, 1 KV head/device) — native 262k context
   - Sweep across context lengths (1k–480k) and batch sizes (1, 4, 32)

## Test Result

All 12 unit/integration tests pass. Benchmark results below.

### Llama 3.1 8B (TP=8, 1 KV head/device) — Median ITL (ms), BS=1

Heuristic resolves to 132 splits vs baseline 32.

| Context | Baseline (32) | Heuristic (132) | Improvement |
|---:|---:|---:|---:|
| 4,096 | 2.577 | 2.874 | -11.5% |
| 16,384 | 2.725 | 2.948 | -8.2% |
| 32,768 | 2.907 | 3.061 | -5.3% |
| 120,000 | 3.689 | 3.556 | +3.6% |
| 256,000 | 4.883 | 4.359 | **+10.7%** |
| 480,000 | 6.836 | 5.619 | **+17.8%** |

Short-context regression is attributed to FA3 combine kernel overhead: 132 splits over a short KV sequence means each split processes very few tokens, and the cost of merging dominates. This is why we gate on `max_model_len > 100k`.

### Qwen3.5-35B-A3B MoE (TP=2, 1 KV head/device) — Median ITL (ms), BS=1

Heuristic resolves to 132 splits vs baseline 32. Native 262k context support.

| Context | Baseline (32) | Heuristic (132) | Improvement |
|---:|---:|---:|---:|
| 1,000 | 5.061 | 5.076 | -0.3% |
| 32,768 | 5.302 | 5.297 | +0.1% |
| 120,000 | 5.898 | 5.670 | **+3.9%** |
| 250,000 | 6.743 | 6.122 | **+9.2%** |

No short-context regression — the MoE model's higher baseline latency makes the combine kernel overhead negligible.

### Higher batch sizes (Llama, 120k context) — Median ITL (ms)

| BS | Heuristic Splits | Baseline | Heuristic | Improvement |
|---:|---:|---:|---:|---:|
| 1 | 132 | 3.689 | 3.556 | +3.6% |
| 4 | 33 | 5.755 | 5.766 | -0.2% |
| 32 | 32 | 117.504 | 118.259 | -0.6% |

No regression — at higher batch sizes the heuristic resolves to 33 or 32 splits, matching baseline.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/config/vllm.py [hunk 1]
```diff
@@ -649,6 +649,38 @@ def _post_init_kv_transfer_config(self) -> None:
...
+            _FA3_DEFAULT_NUM_SPLITS,
+            _FA3_HEURISTIC_MIN_CONTEXT,
+        )
+        from vllm.platforms import current_platform
+
+        if (
+            self.model_config is not None
+            and current_platform.is_cuda()
+            and current_platform.is_device_capability(90)
+            and self.model_config.max_model_len > _FA3_HEURISTIC_MIN_CONTEXT
```
Comment (LucasWilkinson) [inline vllm/config/vllm.py]: is `and self.model_config.max_model_len > _FA3_HEURISTIC_MIN_CONTEXT` needed? does it actually hurt perf to set it higher?

why gate on `and current_platform.is_device_capability(90)`? I think this could help ampere and blackwell

File: vllm/config/vllm.py [hunk 2]
```diff
@@ -1019,6 +1051,20 @@ def has_blocked_weights():
...
         current_platform.check_and_update_config(self)
 
+        # Resolve flash_attn_max_num_splits_for_cuda_graph if not set by user.
+        if self.attention_config.flash_attn_max_num_splits_for_cuda_graph is None:
+            from vllm.config.attention import _FA3_DEFAULT_NUM_SPLITS
+
+            resolved = self._resolve_fa_num_splits()
+            self.attention_config.flash_attn_max_num_splits_for_cuda_graph = resolved
+            if resolved != _FA3_DEFAULT_NUM_SPLITS:
+                logger.info(
```
Comment (LucasWilkinson) [inline vllm/config/vllm.py]: nit: I think this can be debug; I think this excessive for most users

File: vllm/config/vllm.py [hunk 3]
```diff
@@ -649,6 +649,38 @@ def _post_init_kv_transfer_config(self) -> None:
         # This is the same for all backends
         self.kv_transfer_config.kv_role = "kv_both"
 
+    def _resolve_fa_num_splits(self) -> int:
```
Comment (LucasWilkinson) [inline vllm/config/vllm.py]: nit: i think this should be done in the FA backend to avoid polluting `vllm/config/vllm.py` with backend specific stuff

File: vllm/config/vllm.py [hunk 4]
```diff
@@ -1019,6 +1051,20 @@ def has_blocked_weights():
             )
         current_platform.check_and_update_config(self)
 
+        # Resolve flash_attn_max_num_splits_for_cuda_graph if not set by user.
+        if self.attention_config.flash_attn_max_num_splits_for_cuda_graph is None:
```
Comment (LucasWilkinson) [inline vllm/config/vllm.py]: nit: i think this should be done in the FA backend to avoid polluting `vllm/config/vllm.py` with backend specific stuff

**Review by LucasWilkinson — COMMENTED** [review]
I like this idea! thanks for the contribution; left some comments

File: vllm/config/vllm.py (hunk 1 above)
Comment (jmkuebler) [reply vllm/config/vllm.py]: > is and self.model_config.max_model_len > _FA3_HEURISTIC_MIN_CONTEXT needed? does it actually hurt perf to set it higher?

I just observed the the combine overhead is larger than the gains when the context is shorter than around 100k.

> why gate on and current_platform.is_device_capability(90)? I think this could help ampere and blackwell

I gated on SM90 because FA3 is the only one that uses the AOT scheduling and thus the parameter. I guess it would be fine to remove this check though here...

**Comment by jmkuebler** [comment]
@LucasWilkinson thanks a lot for your comments and interest to merge this. I moved the functions according to your suggestions.

But I think this entire setting only applies to FA3 on Hopper. Could you take another look please?
